File: id/id/src/main/java/io/piano/android/id/PianoIdActivity.java
package io.piano.android.id;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.widget.ContentLoadingProgressBar;

import io.piano.android.id.models.PianoIdToken;
import timber.log.Timber;

public class PianoIdActivity extends AppCompatActivity implements PianoIdJsInterface {

    static final String KEY_WIDGET = "io.piano.android.id.PianoIdActivity.WIDGET";
    static final String KEY_DISABLE_SIGN_UP = "io.piano.android.id.PianoIdActivity.DISABLE_SIGN_UP";

    static final String JS_INTERFACE_NAME = "PianoIDMobileSDK";

    static final int OAUTH_PROVIDER_REQUEST_CODE = 1;

    @VisibleForTesting
    WebView webView;
    private ContentLoadingProgressBar progressBar;
    private String widget;
    private boolean disableSignUp;
    private PianoIdClient client;
    private PianoIdJsInterface jsInterface;

    static Intent buildIntent(Context context, boolean disableSignUp, String widget) {
        return new Intent(context, PianoIdActivity.class)
                .putExtra(KEY_DISABLE_SIGN_UP, disableSignUp)
                .putExtra(KEY_WIDGET, widget)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.id);
        client = PianoId.getClient();
        jsInterface = new PianoIdJavascriptDelegate(this);
        webView = findViewById(R.id.webview);
        progressBar = findViewById(R.id.progressBar);
        processIntent(getIntent());

        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                super.onProgressChanged(view, newProgress);
                progressBar.setProgress(newProgress);
            }
        });
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
                progressBar.show();
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                progressBar.hide();
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                Timber.d(url);
                progressBar.show();
                return false;
            }
        });
        client.getSignInUrl(disableSignUp, widget, new PianoIdCallback<String>() {
            @Override
            public void onSuccess(String url) {
                progressBar.setIndeterminate(false);

                webView.addJavascriptInterface(jsInterface, "PianoIDMobileSDK");
                webView.clearHistory();
                webView.loadUrl(url);
            }

            @Override
            public void onFailure(PianoIdException exception) {
                setFailureResultData(exception);
            }
        });
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        processIntent(intent);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == OAUTH_PROVIDER_REQUEST_CODE) {
            if (data == null) {
                setFailureResultData(new IllegalStateException("Result intent is null"));
                return;
            }
            try {
                switch (resultCode) {
                    case Activity.RESULT_OK:
                        String provider = data.getStringExtra(PianoId.KEY_OAUTH_PROVIDER_NAME);
                        String token = data.getStringExtra(PianoId.KEY_OAUTH_PROVIDER_TOKEN);
                        if (provider == null || token == null) {
                            setFailureResultData(new IllegalArgumentException("Result intent should provide provider name and token"));
                            return;
                        }
                        String code = client.buildResultJsCommand(provider, token);
                        evaluateJavascript(code);
                        return;
                    case Activity.RESULT_CANCELED:
                        setResult(Activity.RESULT_CANCELED);
                        finish();
                        return;
                    case PianoId.RESULT_ERROR:
                        int statusCode = data.getIntExtra(PianoId.KEY_ERROR, 0);
                        setFailureResultData(client.getStoredException(statusCode));
                }
            } catch (Exception e) {
                setFailureResultData(e);
            }
        } else super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        webView.saveState(outState);
    }

    @Override
    protected void onDestroy() {
        webView.removeJavascriptInterface(JS_INTERFACE_NAME);
        jsInterface = null;
        super.onDestroy();
    }

    void processIntent(Intent intent) {
        widget = intent.getStringExtra(KEY_WIDGET);
        disableSignUp = intent.getBooleanExtra(KEY_DISABLE_SIGN_UP, false);
    }

    void setSuccessResultData(PianoIdToken token) {
        Intent intent = new Intent();
        intent.putExtra(PianoId.KEY_TOKEN, token);
        setResult(Activity.RESULT_OK, intent);
        PianoIdCallback<PianoIdToken> tokenCallback = client.getTokenCallback();
        if (tokenCallback != null)
            tokenCallback.onSuccess(token);
        finish();
    }

    void setFailureResultData(Throwable throwable) {
        Intent intent = new Intent();
        PianoIdException exception = Utils.wrapException(throwable);
        intent.putExtra(PianoId.KEY_ERROR, client.saveException(exception));
        setResult(PianoId.RESULT_ERROR, intent);
        PianoIdCallback<PianoIdToken> tokenCallback = client.getTokenCallback();
        if (tokenCallback != null)
            tokenCallback.onFailure(exception);
        finish();
    }

    void evaluateJavascript(@NonNull String code) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webView.evaluateJavascript(code, null);
        } else {
            webView.loadUrl("javascript:" + code);
        }
    }

    @Override
    public void socialLogin(String payload) {
        try {
            Intent intent = client.buildSocialAuthIntent(this, payload);
            startActivityForResult(intent, OAUTH_PROVIDER_REQUEST_CODE);
        } catch (Exception e) {
            setFailureResultData(e);
        }
    }

    @Override
    public void loginSuccess(String payload) {
        try {
            PianoIdToken token = client.buildToken(payload);
            setSuccessResultData(token);
        } catch (Exception e) {
            setFailureResultData(e);
        }
    }

    @Override
    public void cancel() {
        setResult(Activity.RESULT_CANCELED);
        finish();
    }
}

File: id/id/src/main/java/io/piano/android/id/PianoIdActivity.kt
package io.piano.android.id

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Message
import android.webkit.CookieManager
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.OnBackPressedCallback
import androidx.annotation.RestrictTo
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import io.piano.android.id.PianoId.Companion.isPianoIdUri
import io.piano.android.id.PianoIdClient.Companion.toPianoIdException
import io.piano.android.id.databinding.ActivityPianoIdBinding
import io.piano.android.id.models.OAuthFailureResult
import io.piano.android.id.models.OAuthSuccessResult
import io.piano.android.id.models.PianoIdAuthResult
import io.piano.android.id.models.PianoIdToken

class PianoIdActivity : AppCompatActivity(), PianoIdJsInterface {
    @VisibleForTesting
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    private lateinit var binding: ActivityPianoIdBinding
    private val client: PianoIdClient = PianoId.getInstance()

    private val jsInterface = PianoIdJavascriptDelegate(this, client.javascriptInterface)
    private var widget: String? = null
    private var disableSignUp: Boolean = false
    private var stage: String? = null

    private val oauthResult = registerForActivityResult(OAuthResultContract()) {
        when (it) {
            null -> {
                setResult(Activity.RESULT_CANCELED)
                finish()
            }

            is OAuthSuccessResult -> binding.webview.evaluateJavascript(it.jsCommand, null)
            is OAuthFailureResult -> setFailureResultData(it.exception)
        }
    }

    private val webviewBackPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            binding.webview.goBack()
        }
    }

    init {
        addOnNewIntentListener {
            it?.process()
        }
    }

    @SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPianoIdBinding.inflate(layoutInflater)
        setContentView(binding.root)
        intent?.process()
        onBackPressedDispatcher.addCallback(webviewBackPressedCallback)
        with(binding) {
            webview.apply {
                savedInstanceState?.let { restoreState(it) }
                settings.javaScriptEnabled = true
                settings.setSupportMultipleWindows(true)
                webChromeClient = object : WebChromeClient() {
                    override fun onProgressChanged(view: WebView, newProgress: Int) {
                        super.onProgressChanged(view, newProgress)
                        progressBar.progress = newProgress
                    }

                    override fun onCreateWindow(
                        view: WebView,
                        isDialog: Boolean,
                        isUserGesture: Boolean,
                        resultMsg: Message,
                    ): Boolean {
                        if (!isUserGesture) {
                            return false
                        }
                        with(resultMsg) {
                            (obj as WebView.WebViewTransport).webView = WebView(view.context)
                            sendToTarget()
                        }
                        return true
                    }
                }
                webViewClient = object : WebViewClient() {
                    override fun onPageStarted(view: WebView, url: String?, favicon: Bitmap?) {
                        super.onPageStarted(view, url, favicon)
                        webviewBackPressedCallback.isEnabled = view.canGoBack()
                        progressBar.show()
                    }

                    override fun onPageFinished(view: WebView, url: String?) {
                        super.onPageFinished(view, url)
                        evaluateJavascript(FormHelper.buildConsentsCode(), null)
                        progressBar.hide()
                    }

                    @Deprecated("Deprecated in Java")
                    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                        val isPianoIdRedirectUrl = Uri.parse(url).isPianoIdUri()
                        if (isPianoIdRedirectUrl) {
                            setFailureResultData(
                                IllegalStateException("User already authorized, call signOut before login")
                            )
                        }
                        progressBar.show()
                        return isPianoIdRedirectUrl
                    }
                }
            }
            val url = client.getSignInUrl(disableSignUp, widget, stage)
            CookieManager.getInstance().setCookie(url, "${client.aid}__ut=")
            progressBar.isIndeterminate = false
            webview.apply {
                addJavascriptInterface(jsInterface, JS_INTERFACE_NAME)
                clearCache(true)
                clearHistory()
                loadUrl(url)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        binding.webview.saveState(outState)
    }

    override fun onDestroy() {
        binding.webview.removeJavascriptInterface(JS_INTERFACE_NAME)
        super.onDestroy()
    }

    override fun socialLogin(payload: String?) {
        runCatching {
            oauthResult.launch(requireNotNull(payload))
        }.onFailure {
            setFailureResultData(it)
        }
    }

    override fun registerSuccess(payload: String?) = parsePayload(payload, true)

    override fun loginSuccess(payload: String?) = parsePayload(payload, false)

    override fun error(payload: String?) = setFailureResultData(client.parseJsError(payload))

    override fun cancel() {
        setResult(Activity.RESULT_CANCELED)
        finish()
    }

    private fun parsePayload(payload: String?, isNewUser: Boolean) {
        runCatching {
            setSuccessResultData(client.buildToken(requireNotNull(payload)), isNewUser)
        }.onFailure {
            setFailureResultData(it)
        }
    }

    private fun setSuccessResultData(token: PianoIdToken, isNewUser: Boolean) {
        setResult(
            Activity.RESULT_OK,
            Intent().putExtra(PianoId.KEY_TOKEN, token).putExtra(PianoId.KEY_IS_NEW_USER, isNewUser)
        )
        client.authCallback?.invoke(PianoIdAuthResult.success(token, isNewUser))
        finish()
    }

    private fun setFailureResultData(throwable: Throwable) {
        val exc = throwable.toPianoIdException()
        setResult(PianoId.RESULT_ERROR, Intent().putExtra(PianoId.KEY_ERROR, client.saveException(exc)))
        client.authCallback?.invoke(PianoIdAuthResult.failure(exc))
        finish()
    }

    internal fun Intent.process() {
        widget = getStringExtra(KEY_WIDGET)
        disableSignUp = getBooleanExtra(KEY_DISABLE_SIGN_UP, false)
        stage = getStringExtra(KEY_STAGE)
    }

    companion object {
        internal const val KEY_WIDGET = "io.piano.android.id.PianoIdActivity.WIDGET"
        internal const val KEY_DISABLE_SIGN_UP = "io.piano.android.id.PianoIdActivity.DISABLE_SIGN_UP"
        internal const val KEY_STAGE = "io.piano.android.id.PianoIdActivity.STAGE"
        internal const val JS_INTERFACE_NAME = "PianoIDMobileSDK"

        @JvmStatic
        internal fun buildIntent(context: Context, disableSignUp: Boolean, widget: String?, stage: String?): Intent =
            Intent(context, PianoIdActivity::class.java)
                .putExtra(KEY_DISABLE_SIGN_UP, disableSignUp)
                .putExtra(KEY_WIDGET, widget)
                .putExtra(KEY_STAGE, stage)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
    }
}
